### ISSUE TITLE
[FR] TMC5161 stepper driver support

### ISSUE THREAD

**Issue by LT1010** [MarlinFirmware/Marlin#16365]
I have encountered a problem, I use the latest marlin firmware and 5160 is OK on my motherboard. The configuration is not changed, only the drive is changed to 5161, and the information displayed using M122 is normal, but when I move the XYZ axis, there is only a little movement and then it cannot move. I used M122 to see the information changed.
move_before:
![move_before](https://user-images.githubusercontent.com/56394552/71538271-06c07a00-2963-11ea-8c81-664d655532f4.png)
move_x_asis:
![move_x_axis](https://user-images.githubusercontent.com/56394552/71538276-29529300-2963-11ea-96df-1d3bbb727281.png)

**Comment by LT1010** [comment]
Please tell me where the problem is(TMC5160 is OK on my motherboard).

**Comment by sjasonsmith** [comment]
Can you add your configuration files, including your pins file if you have modified it.

**Comment by sjasonsmith** [comment]
I would also like to know if you see the same results every time you run M122, or if the results are somewhat random.

**Comment by LT1010** [comment]
> 我还想知道每次运行M122时是否看到相同的结果，或者结果是否有些随机。

Hi,this is my configuration file, the pin file has not been moved
[configuration.zip](https://github.com/MarlinFirmware/Marlin/files/4006675/configuration.zip).
Every time I move any axis, the mstep of the corresponding axis will become 256, and then the motor will not move.

**Comment by sjasonsmith** [comment]
This might be related to https://github.com/MarlinFirmware/Marlin/issues/16311.

That issue was closed since the fix belongs to TMCStepper, but there is a workaround described that you could try.

**Comment by LT1010** [comment]
> This might be related to #16311.
> 
> That issue was closed since the fix belongs to TMCStepper, but there is a workaround described that you could try.

Hello, is there any other solution besides this, I will try one by one.

**Comment by LT1010** [comment]
I used a logic analyzer and here is the result：
![TMC5160](https://user-images.githubusercontent.com/56394552/71541670-9bdd6600-2997-11ea-85c2-5a19a5a15ec5.png)
![TMC5161](https://user-images.githubusercontent.com/56394552/71541671-9c75fc80-2997-11ea-890b-e5e59a310ffd.png)

**Comment by LT1010** [comment]
My problem has not been solved, this seems to have nothing to do with TMCStepper, I try to reduce the speed of SW_SPI, and the result of M122 is not random, it only changes after I move the axis, and each time after the change, M122 The results are also fixed. I also want to know if there are other ways to get my 5161 working properly.

**Comment by lijinlin19970813** [comment]
> My problem has not been solved, this seems to have nothing to do with TMCStepper, I try to reduce the speed of SW_SPI, and the result of M122 is not random, it only changes after I move the axis, and each time after the change, M122 The results are also fixed. I also want to know if there are other ways to get my 5161 working properly.

![CMP](https://user-images.githubusercontent.com/56394552/71565708-6b5b1080-2aec-11ea-838f-6c4eab870c22.png)

**Comment by LT1010** [comment]
> My problem has not been solved, this seems to have nothing to do with TMCStepper, I try to reduce the speed of SW_SPI, and the result of M122 is not random, it only changes after I move the axis, and each time after the change, M122 The results are also fixed. I also want to know if there are other ways to get my 5161 working properly.

![CMP](https://user-images.githubusercontent.com/56394552/71565750-be34c800-2aec-11ea-9fc4-acf42f6c5057.png)

**Comment by boelle** [comment]
@sjasonsmith does the info from @lijinlin19970813 make any bells ring?

**Comment by sjasonsmith** [comment]
> @sjasonsmith does the info from @lijinlin19970813 make any bells ring?

@boelle These are SPI-connected drivers. Everything I have been has been for UART-connected drivers. I've never used either of these.

**Comment by Bob-the-Kuhn** [comment]
The behavior you're describing could be due to too little power being used for the speed, acceleration, jerk, and microstep settings.

One possibility is the RSENSE settings specified in the Configuration_adv.h file are wrong.  The current settings are 0.11 Ohms.  A picture I saw that showed both the 5160 & 5161 pololu driver modules showed 0.075 Ohm resistors on the 5160 and 0.062 Ohm resistors on the 5161.  You'll need to physically examine  your modules to see what resistors are actually used.

I doubt that this is your problem.  If my numbers are accurate you are actually sending more current to the steppers than specified in Configuration_adv.h.

I suggest you try some **RIDICULOUSLY** conservative settings to see if that is the problem.  Try setting microsteps to 4, cut steps per unit by 4, cut acceration max by 10, cut max speed by 10 and decrease jerk by 10.  Then manually issue really slow movement commands (G0 X20 F200).

**Comment by lijinlin19970813** [comment]
> The behavior you're describing could be due to too little power being used for the speed, acceleration, jerk, and microstep settings.
> 
> One possibility is the RSENSE settings specified in the Configuration_adv.h file are wrong. The current settings are 0.11 Ohms. A picture I saw that showed both the 5160 & 5161 pololu driver modules showed 0.075 Ohm resistors on the 5160 and 0.062 Ohm resistors on the 5161. You'll need to physically examine your modules to see what resistors are actually used.
> 
> I doubt that this is your problem. If my numbers are accurate you are actually sending more current to the steppers than specified in Configuration_adv.h.
> 
> I suggest you try some **RIDICULOUSLY** conservative settings to see if that is the problem. Try setting microsteps to 4, cut steps per unit by 4, cut acceration max by 10, cut max speed by 10 and decrease jerk by 10. Then manually issue really slow movement commands (G0 X20 F200).
I changed the RSENSE to over 0.062, it is still invalid, but I will try what you said to change other parameters to try the effect.

**Comment by LT1010** [comment]
> The behavior you're describing could be due to too little power being used for the speed, acceleration, jerk, and microstep settings.
> 
> One possibility is the RSENSE settings specified in the Configuration_adv.h file are wrong. The current settings are 0.11 Ohms. A picture I saw that showed both the 5160 & 5161 pololu driver modules showed 0.075 Ohm resistors on the 5160 and 0.062 Ohm resistors on the 5161. You'll need to physically examine your modules to see what resistors are actually used.
> 
> I doubt that this is your problem. If my numbers are accurate you are actually sending more current to the steppers than specified in Configuration_adv.h.
> 
> I suggest you try some **RIDICULOUSLY** conservative settings to see if that is the problem. Try setting microsteps to 4, cut steps per unit by 4, cut acceration max by 10, cut max speed by 10 and decrease jerk by 10. Then manually issue really slow movement commands (G0 X20 F200).

I changed the RSENSE to over 0.062, it is still invalid, but I will try what you said to change other parameters to try the effect.

**Comment by Bob-the-Kuhn** [comment]
Also double the current settings in configuration_adv.h

**Comment by boelle** [comment]
@lijinlin19970813 did the last suggestion from @Bob-the-Kuhn help?

do we have a bug or does it more look like a config error?

**Comment by XBrav** [comment]
It still happens, at least for me. With the latest TMCStepper library (0.6.1 as of posting), I can get up and down movement by jogging manually. Homing seems to work fine too, however, around the first few probes, the Z simply stops moving.

Whatever is happening, the stepper entirely seems to shut off. I can freely move the rod the moment the failure occurs. Unfortunately, I'm using a BTT E3 board, and I don't have enough memory for TMC_DEBUG currently.

**Comment by XBrav** [comment]
Ok, I stripped down the code enough to get M122 on the machine. It's an E3 DIP, with 2208s in UART on the X, Y, and E0. Z is the 5161.
```
SENDING:M122
		X	Y	Z	E
Enabled		false	false	false	false
Set current	650	800	1000	700
RMS current	642	795	990	673
MAX current	905	1121	1396	949
Run current	20/31	25/31	19/31	21/31
Hold current	18/31	22/31	17/31	18/31
Global scaler			133/256
CS actual	18/31	22/31	17/31	18/31
PWM scale	21	25	16	21
vsense		1=.18	1=.18		1=.18
stealthChop	true	true	true	true
msteps		16	16	16	16
tstep		max	max	max	max
pwm
threshold
[mm/s]
OT prewarn	false	false	false	false
OT prewarn has
been triggered	false	false	false	false
off time	3	3	3	3
blank time	24	24	24	24
hysteresis
-end		-1	-1	-1	-1
-start		1	1	1	1
Stallguard thrs			0
DRVSTATUS	X	Y	Z	E
stallguard
sg_result			0
fsactive
stst		*	*	*	*
olb
ola
s2gb
s2ga
otpw
ot
157C
150C
143C
120C
s2vsa
s2vsb
Driver registers:
		X	0xC0:12:00:00
		Y	0xC0:16:00:00
		Z	0x80:11:40:00
		E	0xC0:12:00:00
Testing X connection... OK
Testing Y connection... OK
Testing Z connection... OK
Testing E connection... OK`
```
I then perform a move (up is fine, it's the direction change too that causes the issue), the microsteps changes to 256 and the address changes as well:
```
G1 Z10
SENDING:G1 Z10
G1 Z0
SENDING:G1 Z0
M122
SENDING:M122
		X	Y	Z	E
Enabled		false	false	false	false
Set current	650	800	1000	700
RMS current	642	795	990	673
MAX current	905	1121	1396	949
Run current	20/31	25/31	19/31	21/31
Hold current	18/31	22/31	17/31	18/31
Global scaler			133/256
CS actual	18/31	22/31	8/31	18/31
PWM scale	21	25	8	21
vsense		1=.18	1=.18		1=.18
stealthChop	true	true	false	true
msteps		16	16	256	16
tstep		max	max	max	max
pwm
threshold
[mm/s]
OT prewarn	false	false	false	false
OT prewarn has
been triggered	false	false	false	false
off time	3	3	0	3
blank time	24	24	36	24
hysteresis
-end		-1	-1	-1	-1
-start		1	1	6	1
Stallguard thrs			0
DRVSTATUS	X	Y	Z	E
stallguard
sg_result			0
fsactive
stst		*	*	*	*
olb
ola
s2gb
s2ga
otpw
ot
157C
150C
143C
120C
s2vsa
s2vsb
Driver registers:
		X	0xC0:12:00:00
		Y	0xC0:16:00:00
		Z	0x80:08:00:00
		E	0xC0:12:00:00
Testing X connection... OK
Testing Y connection... OK
Testing Z connection... OK
Testing E connection... OK
```
At this point, the stepper no longer responds, and requires a reboot of the firmware to regain control. Depending on the motion, the Z address will change each time the failure occurs. I've seen 0x80:08:40:00 as well as 0x80:0D:40:00.

[Marlin.zip](https://github.com/MarlinFirmware/Marlin/files/4057370/Marlin.zip)

I have attached my config files for reference.

**Comment by RonnyOs** [comment]
Hi,

I‘ve got the same problem with my TMC5161 drivers. I tried the decrease of sw spi clock with no succes. If i force the drivers to use chop mode, the Steppers are working. Current Must be in a small Range window to prevent it from stop. If I try to use faster movements the Steppers stop working at Point of Switch to spread Cycle.

**Comment by boelle** [comment]
a lot of updates and fixing has happend in the last week, is the problem still there?

**Comment by XBrav** [comment]
@boelle I'll try and set up a test rig tonight. https://github.com/teemuatlut/TMCStepper/blob/master/src/source/TMC_platforms.h has not been modified for the stm32 platform, only for the LPC176x family.

Luckily, I have one E3 DIP machine standing by.

**Comment by XBrav** [comment]
Just compiled 2.0.3.

Environment:
BIGTREETECH E3 DIP 1.0 (STM32F103RC, 256k, NOUSB)
Steppers:
X: BIGTREETECH TMC2208 V3.0 UART
Y: BIGTREETECH TMC2208 V3.0 UART
Z: BIGTREETECH TM5161 V1.1 SPI (Configured as TMC5160 with TMC_USE_SW_SPI)
E: BIGTREETECH TMC2208 V3.0 UART

All steppers detected. Issuing a home command, Z engages but does not move. X and Y homing continues, but Z is seemingly disengaged. Unable to enable M122 due to lack of memory. But likely previous issues noted above.

Without recompiling the firmware, I replaced the problematic stepper with a BIGTREETECH TMC5160 V1.2. As expected, the stepper engages and moves in both forward and reverse directions without issue.

It appears as though there were further changes to the 5161 series than simply integrating the MOSFETs into the die. Adding to the concerns, the product page notes product changes are currently underway: https://www.trinamic.com/products/integrated-circuits/details/tmc5161/

I will reach out to Trinamic for comment. For the time being, I would recommend users stick with the refined 5160s, as they are operating as expected.

**Comment by boelle** [comment]
is it the same with bugfix 2.0.x ?

**Comment by sjasonsmith** [comment]
> is it the same with bugfix 2.0.x ?

bf2 is only one commit ahead of 2.0.3. Testing with it would be a waste of time.

**Comment by einencool** [comment]
I still got the same issue, I posted it on the BTT Github, but nobody responsed to it in the last 5 weeks...

 I also cut the CLK-Pin from the 5160 and the 5161.
The 5160 are working fine, but the 5161 make a little move and stop after that without any failure.
I changed the "RSense" from 0.075 (5160) to 0.060 (5161)

I tried it on the SKR Pro V1.1 and today with a completly new setup on the SKR V1.4 Turbo
The 5160 work well and the 5161 are stopping immediatly.

Can someone tell me if it is normal, that the drivers are very loud with 0.9° Steppers? 
I'm using them in a CoreXY and need around 1.5V to get the Steppers warm up to 40°C . If I use only 1.2V the Steppers are not so noisy, but the Steppers are cold as ice and I'm loosing steps like hell...

**Comment by Malaquitte** [comment]
These drivers are really crappy. I tried everything on a SKR 1.1 Pro, I couldn't get them to work.
I don't understand how Bigtreetech makes them work on this video https://www.youtube.com/watch?v=UW_JtHcgcd4, support never answered my questions!

**Comment by Werebear93** [comment]
I can't seem to get my TMC5160 working, either. Connected one as a test to E0 and M122 does the following:

![m122_result](https://user-images.githubusercontent.com/46569697/73751783-0f754e80-4760-11ea-9b2c-78dab06a56a7.jpg)


I've snipped the CLK pin, tried with and without connecting CLK to ground, adjusted RSENSE to 0.075 Ohm and nothing seems to work. 
Here's my current configuration (nightly bugfix build, 02/04/2020): 
[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/4153806/Configuration.zip)

Maybe someone has an idea of what I could have missed. I haven't tried limiting the SPI speed and honestly I don't really know how to do that, either.

**Comment by thinkyhead** [comment]
Well, for one thing, we don't formally support the 5161. As for the 5160 not working, well I'm not sure how well supported it is now in TMCStepper either. It is all very new, so we may have to wait on @teemuatlut to find some free time to troubleshoot them generally. We have a set of them here, but at the moment my custodial duties are taking up too much time to dive into testing. But perhaps @shitcreek and I will find some time to play around with 5160's later this week.

**Comment by Malaquitte** [comment]
If it's only a matter of weeks, I'm still hopeful. With my knowledge of steppers drivers I can unfortunately not help. Thanks for everything else.

**Comment by 2kingsrock** [comment]
@Werebear93 I dont know what physical driver slots are configured on your board, but this looks off to me.  Z/Z2 if you are using dual Z motors this would be strange.
```cpp
  #if AXIS_IS_TMC(Z)
    #define Z_CURRENT       800
    #define Z_CURRENT_HOME  Z_CURRENT
    #define Z_MICROSTEPS     16
    #define Z_RSENSE          0.11
    #define Z_CHAIN_POS      -1
  #endif

  #if AXIS_IS_TMC(Z2)
    #define Z2_CURRENT      800
    #define Z2_CURRENT_HOME Z2_CURRENT
    #define Z2_MICROSTEPS    16
    #define Z2_RSENSE         0.075
    #define Z2_CHAIN_POS     -1
  #endif
```

Also checking your voltage, are you running a 24v or 12v system?
```cpp
#define CHOPPER_TIMING CHOPPER_DEFAULT_12V
```

**Comment by GeminiServer** [comment]
Have exactly the same issue. After disabling SOFTWARE_DRIVER_ENABLE and STEALTHCHOP_X _Y _Z _E, the TMC5161 runs. But no features then... TMC5161 on SKR_1.4_Pro with SPI and recent Marlin Bugfix branch.

RSENSE - Nothing helped... here are the Informations:
```
// | --------------------------------------------------------------------------|
// | RSENSE [Ω]  | RMS current [A](CS=31) | Sine wave peak current [A] (CS=31) |
// | 0.22        | 1.1                    | 1.5                                |
// | 0.15        | 1.6                    | 2.2                                |
// | 0.12        | 2.0                    | 2.8                                |
// | 0.10        | 2.3                    | 3.3                                |
// | 0.075       | 3.1                    | 4.4                                |
// | 0.066       | 3.5                    | 5.0                                |
// | 0.060       | 3.8                    | 5.4                                | <-- 
// | --------------------------------------------------------------------------|
// Documentation --> 9 Selecting Sense Resistors List - 
```
https://www.trinamic.com/fileadmin/assets/Products/ICs_Documents/TMC5161-datasheet_Rev1.02.pdf

**Comment by thinkyhead** [comment]
@GeminiServer — Is there any recent or old Marlin version where the driver works ok?

**Comment by GeminiServer** [comment]
> @GeminiServer — Is there any recent or old Marlin version where the driver works ok?

Nope! I gone stepply back to till 2.0.1 and no luck.

I spend to much time on this drivers. I'll send these crappy drivers back and order the TMC5160, which seems to working and supported better than the 5161! The only difference in 5161, that they have the embedded mosfets... so don't really makes it attractive.  And the support of BTT is ZERO, so i suggest all to do the same... throw them away and change to the 5160!

**Comment by GeminiServer** [comment]
But just fyi - The RSENSE on the modules seems to be different then the sheets:
This is TMC5161 v1.1 with RSENSE of 0.062! Which seems Initial to work, but stops after a while, like also described above from other users. 

![image](https://user-images.githubusercontent.com/7021180/74226001-0afae980-4cbc-11ea-9103-68e558510810.png)

**Comment by thinkyhead** [comment]
Thanks for the very useful info. Hopefully this will be sorted in the hardware. If there's any special initialization needed for these drivers, we'd have to wait for TMCStepper to bring that in. I'll have a peek at the datasheet soon and see if anything stands out in contrast to the 5160.

**Comment by viktoschi** [comment]
tested the driver yesterday with arduino mega and the examples in TMCStepper Repo. Stepper drivers worked fine.

**Comment by GeminiServer** [comment]
**Update:** 
I was now able to get the TMC5161 run with recent Marlin Bugfix 2.0.x. After checking latest changes at the TMCStepper-Library, i think that i found the reason and i did a test run for a couple of hours now. Till now, the driver and stepper runs without any issues. 

**Hardware setup:**: BTT TMC5161 v1.1 and SKR 1.4 Turbo (LPC1769)
**Marlin version:**: Branch - bugfix-2.0.x @ 12b3807a7f443470c290ec8b6fb6eefdc5b39280

**My configuration:**
For BTT TMC5161 **v1.1** with **R=0.062**

// RSENSE is "**0.062f**" is **float**. TMCStepper library v0.6.2 changed it to float! Therefore the float define! https://github.com/teemuatlut/TMCStepper/commit/4016500179171cba0614e116843d77229138af84

Be sure you use the TMCStepper-Lib **v0.6.2** in your platformio.ini lib_debs:
> TMCStepper@>=0.6.2,<1.0.0

The configuration_adv.h:
> #define TMC5161_RSENSE 0.062f   // For X, Y, Z end E
> 
> #define HOLD_MULTIPLIER    0.5
> #define INTERPOLATE              false
> #define XYZE_CURRENT          800
> #define XYZE_MICROSTEPS   16
> 
> #define STEALTHCHOP_XY
> #define STEALTHCHOP_Z
> #define STEALTHCHOP_E
> 
> //#define SOFTWARE_DRIVER_ENABLE  // --> is disabled!

Give it a try and let us know if this also work for others then me. 


Cheers.

**Comment by XBrav** [comment]
@GeminiServer, can you confirm if it works with 2.0.3? I will need to test this later on the STM32 platform. Previously, it was apparently resolved on the LPC176x family due to the 240ns delay imposed on the commands.

**Comment by Malaquitte** [comment]
I did a quick test yesterday on Marlin 2.0.3, motors aren't running.

**Comment by GeminiServer** [comment]
@Malaquitte did you also set the #define INTERPOLATE to false?

**Comment by Malaquitte** [comment]
@GeminiServer I've done everything you said above, taking care to delete the .pio folder to make sure I inherit the right version of TMCStepper.
Did you declare TMC5161 in #define X(Y)(Z)(E)_DRIVER_TYPE ?

**Comment by GeminiServer** [comment]
@Malaquitte yes, i set the driver tpye for xyze to TMC5160.

The TMC5161 driver works only if INTERPOLATE is set to false. If you go higher with the current they do crazy things. Unfortunately i switched now to TMC5160, which works like it should be and i will send the TMC5161 back to BTT (they also have no idea why the TMC5161 not work with Marlin - got mail from their dev support). As i said before, the only difference that i see are the embedded mosfets in 5161.

**Comment by einencool** [comment]
Today I tried to get my 5161 V1.1 to work, but they still move a little bit and then they stop.
I downloaded a fresh Marlin 2.0.x Bugfix and changed the TMC lib to 0.6.2

This is the following Config:
```cpp
#define X_DRIVER_TYPE  TMC5160
#define Y_DRIVER_TYPE  TMC5160
#define Z_DRIVER_TYPE  TMC2209
#define E0_DRIVER_TYPE TMC2209

  #define HOLD_MULTIPLIER    0.5  // Scales down the holding current from run current
  #define INTERPOLATE       false  // Interpolate X/Y/Z_MICROSTEPS to 256

  #if AXIS_IS_TMC(X)
    #define X_CURRENT       1400        // (mA) RMS current. Multiply by 1.414 for peak current.
    #define X_CURRENT_HOME  X_CURRENT  // (mA) RMS current for sensorless homing
    #define X_MICROSTEPS     16    // 0..256
    #define X_RSENSE          0.062f
    #define X_CHAIN_POS      -1    // <=0 : Not chained. 1 : MCU MOSI connected. 2 : Next in chain, ...
  #endif

  #define TMC_USE_SW_SPI
  //#define SOFTWARE_DRIVER_ENABLE
  #define STEALTHCHOP_XY
  #define CHOPPER_TIMING CHOPPER_DEFAULT_24V
  #define HYBRID_THRESHOLD
  #define TMC_DEBUG
```
So I don't know, what to do next... I've cut the "CLK Pin" like on the 5160 V1.2 mentioned from the BTT support. Well, I think I've wasted 65 € with this pieces...

**Comment by viktoschi** [comment]
i had success with lower currents.
but still have problems with the current on dual z motors with one stepper driver

**Comment by einencool** [comment]
When I lower the current, I've a lot of skipping steps, so that's no good idea for me.
In my case the motors are cold (around 35°C) until I raise the voltage to 1.400V 
With 1.600V the motors are absolutely fine with the TMC5160 V1.2, but they are very loud, like the old A4988. If I lower the voltage they are much quieter, but with 1.200 V I'm only skipping steps and with 0.8V I even can't home the machine. It's a big CoreXY Printer with 0.9° Motors

**Comment by viktoschi** [comment]
got all axis working. had wrong accelerations and axis steps.
its also working with #define INTERPOLATE true. but then you need one driver per motor.
other settings in conf_adv
```cpp
#define X_CURRENT       800  
```
config:
```cpp
#define DEFAULT_AXIS_STEPS_PER_UNIT   { 80, 80, 400, 445 }
#define DEFAULT_MAX_FEEDRATE          { 300, 300, 50, 40 }
#define DEFAULT_MAX_ACCELERATION      { 2000, 2000, 100, 10000 }
#define DEFAULT_ACCELERATION          800    // X, Y, Z and E acceleration for printing moves
#define DEFAULT_RETRACT_ACCELERATION  10000    // E acceleration for retracts
#define DEFAULT_TRAVEL_ACCELERATION   2000    // X, Y, Z acceleration for travel (non printing) 
```
https://github.com/LadyViktoria/Marlin/tree/WIP_TMC5161

**Comment by RonnyOs** [comment]
I am not able to compile since i get an errot in function "restore_trinamic_drivers" something with "stepperZ" "stepperY" "stepperX" first defined here. Anyone knows this problem? with older bugfix2.0 the configuration files were working...

**Comment by viktoschi** [comment]
delete .pio folder and maybe /users/.platformio too
make sure you have TMCStepper@>=0.6.2,<1.0.0 in platformio.ini

**Comment by Malaquitte** [comment]
\o/ It's starting to work for me with the following values:
```cpp
#define X_DRIVER_TYPE  TMC5160

#define X_CURRENT       580 
#define X_CURRENT_HOME  X_CURRENT  
#define X_MICROSTEPS     16    
#define X_RSENSE          0.062
#define X_CHAIN_POS      -1    
```
I'm sure I can make it even better by playing with `X_RENSE`.

**Comment by viktoschi** [comment]
i have notice its better to wait a few seconds until the flash drive comes up in windows before connect with pronterface

**Comment by Malaquitte** [comment]
On .1mm steps it works 2 or 3 times and then nothing...

**Comment by viktoschi** [comment]
also changed accelartion and steps per mm ? 
i wish we could calculate the correct values :/ 
i found my values with try and error.
i am using skr1.4 turbo and 4 drivers. on z driver are 2 motors. everything is external and nothing else connected only motherboard, drivers and motors.
i am using 5 of this
https://www.amazon.de/gp/product/B07CZHLKTC

**Comment by thinkyhead** [comment]
@GeminiServer — Can you just go back one more time and do a flash and reset of EEPROM with the RSENSE values defined with no **f** on the end and compare again with the same values but with an **f** on the end to force it to be a `float` (instead of a `double`)…? If you get the same results I will be much more confident that it's a fixable issue.

I mainly want to confirm that the conversion from `double` to `float` is causing the issue because it seems very odd.

_Be sure to delete the `.pio` folder between builds_, because we want to make sure everything links together from the same starting point.

**Comment by RonnyOs** [comment]
For me it is much better now, but the current has to stay in a very limited range. To much mA or to less will stop the steppers ob high acceleration levels. Also if i increase the microsteps the drivers fill stop working earlier.

**Comment by RonnyOs** [comment]
Today I had an interesting effect. I had some collision on one of the axis. On this collision the stepperdriver also stop working, in M122 then I read out the 256 microsteps! It seems that anything like "collision detection" or "stall guard" will stop the drivers from work after this event.

**Comment by thinkyhead** [comment]
There is something called "standstill" which can get triggered by some events, and this shuts down the driver and leaves it in a weird state. I'm not surprised a limited current range is needed. These drivers are challenging and need very good cooling.

**Comment by viktoschi** [comment]
in my case it works now. but one thing is not working 2 motors on one driver. any hints about that?

**Comment by RonnyOs** [comment]
I don't know what was changed but with a pull today it seems to works very well!

**Comment by viktoschi** [comment]
ok this is strange...
got 2 motors on one stepper working.
flash marlin with both motors connected. reset board. send "G1 Z100 F800" motors not working. ->reset
next just unplug ZB while board is running wait a second replug again and they are both working. 
after a reset the motors still work.
after powering the board off you have to unplug and plug the ZB motor again
can someone confirm this? 
iam using 1,5A motors
here are my only changes
https://github.com/LadyViktoria/Marlin/commit/1fdb455ca17fd584c8f1738c4a324c1dfb23ed47

**Comment by thinkyhead** [comment]
What is Trinamic's guidance on connecting multiple motors to a single driver? Can they handle the current load and resistance? Do they get confused by the combined feedback?

**Comment by viktoschi** [comment]
> The TMC5161 can drive one or two DC motors using one coil output per DC motor. Either a torque
limited operation, or a voltage based velocity control with optional torque limit is possible.

https://www.trinamic.com/fileadmin/assets/Products/ICs_Documents/TMC5161A-datasheet_Rev1.05.pdf
page 106 "21 DC Motor or Solenoid"

my english is not well. but 5161 is nearly the same as 5160. diffence is that the mosfets are on the chip. on 5160 drivers they are extern.

**Comment by teemuatlut** [comment]
DC motor literally means a typical traditional DC motor with commutators and all. You can think of the stepper driver as a 2 channel constant current source and for a stepper motor it conveniently keeps adjusting the coil currents in a "sine" wave and the motor magnets try to follow. Then you get motion.
I think you could drive a couple of LEDs with one for all it cares.

The change in TMCStepper with floats was that the compiler could get confused with two constructors and you needed to specify better which type you passed. This does not happen in Marlin as the user doesn't interact directly with the TMCStepper classes, but with a templated TMCMarlin class that inherits from the library. I don't think there should occur such potential confusion in Marlin, and when passing from the constructor to the library, it already has a strongly typed parameter.
The 5161 class in TMCStepper doesn't do anything beyond constructing a 5160 object and as far as I know there's no software side difference between the two models. I'll try to get my hands on a few if Watterott ever makes a stepStick with the IC.

**Comment by viktoschi** [comment]
if you are from germany i can send you a 5161

**Comment by teemuatlut** [comment]
Thanks but I live in Finland. The Watterott github page topic mentions the 5161 so maybe they're still working on it. Or maybe it was cancelled.

**Comment by RonnyOs** [comment]
Hi,

is there any change in the TMC driver section of Marlin in the last 2-3 weeks into the actual bugfix version? With the update today the problems impaired.

With the old version I was able to print but now, after G29 one of the steppers stop working.

**Comment by Automatica-Mat** [comment]
Is there any solution to this problem? I am using the latest TMCStepper libraries and Marlin Bug Fix and am still having the same issues as the people above.

**Comment by Automatica-Mat** [comment]
As the others have shown after only a short move and then stopping a M122 shows msteps changed to 256. It also strange that the vsense is not being reported for the 5161's ?
My configuration is 
x = 5161
y = 5161
z = 2209
e = 2208

SENDING:M122
		X	Y	Z	E
Address				0
Enabled		false	false	false	false
Set current	800	800	800	800
RMS current	787	787	795	795
MAX current	1110	1110	1121	1121
Run current	12/31	12/31	25/31	25/31
Hold current	6/31	6/31	12/31	12/31
Global scaler	135/256	135/256
CS actual	8/31	8/31	12/31	12/31
PWM scale	917526	4063302	15	14
vsense				1=.18	1=.18
stealthChop	true	true	true	true
msteps		256	256	16	16
tstep		max	max	max	max
pwm
threshold
[mm/s]
OT prewarn	false	false	false	false
OT prewarn has
been triggered	false	false	false	false
off time	0	0	4	4
blank time	36	36	24	24
hysteresis
-end		-1	-1	2	2
-start		6	6	1	1
Stallguard thrs	0	0	0
DRVSTATUS	X	Y	Z	E
stallguard
sg_result	0	0
fsactive
stst		*	*	*	*
olb
ola
s2gb
s2ga
otpw
ot
157C
150C
143C
120C
s2vsa
s2vsb
Driver registers:
		X	0x80:08:40:00
		Y	0x80:08:40:00
		Z	0xC0:0C:00:00
		E	0xC0:0C:00:00
Testing X connection... OK
Testing Y connection... OK
Testing Z connection... OK
Testing E connection... OK

**Comment by Malaquitte** [comment]
Any news?  My TMC5161 are bored in my desk drawer...

**Comment by makemerush** [comment]
> drawer

Same here!

**Comment by thinkyhead** [comment]
I have added a "help wanted" tag to increase the chances that someone will appear who has the time and talent to help with this.

**Comment by Acidice1000** [comment]
Same problem this is the result before I move a axis.  SKR 1.4 Turbo and TMC5161 V1.1
I can move X and Y Once a few millimeters but the moment I change direction it stops.
Connecting...
Printer is now online.
>>> M122 S0
SENDING:M122 S0
		X	Y	Z	E
Enabled		false	false	false	false
Set current	640	640	640	940
RMS current	624	624	624	931
MAX current	880	880	880	1313
Run current	9/31	9/31	9/31	14/31
Hold current	4/31	4/31	4/31	7/31
Global scaler	136/256	136/256	136/256	134/256
CS actual	4/31	4/31	4/31	7/31
PWM scale	4	4	4	7
stealthChop	true	true	true	true
msteps		64	64	64	64
tstep		max	max	max	max
PWM thresh.
[mm/s]
OT prewarn	false	false	false	false
triggered
 OTP		false	false	false	false
off time	4	4	4	4
blank time	24	24	24	24
hysteresis
 -end		2	2	2	2
 -start		1	1	1	1
Stallguard thrs	0	0	0	0
uStep count	1018	1018	1018	1018
DRVSTATUS	X	Y	Z	E
sg_result	0	0	0	0
stallguard
fsactive
stst
olb
ola
s2gb
s2ga
otpw
ot
Driver registers:
		X	0x80:04:40:00
		Y	0x80:04:40:00
		Z	0x80:04:40:00
		E	0x80:07:40:00
Testing X connection... OK
Testing Y connection... OK
Testing Z connection... OK
Testing E connection... OK

This is the result after I moved a axis.

>>> M122 S0
SENDING:M122 S0
		X	Y	Z	E
Enabled		false	false	false	false
Set current	640	640	640	940
RMS current	624	624	624	931
MAX current	880	880	880	1313
Run current	9/31	9/31	9/31	14/31
Hold current	4/31	4/31	4/31	7/31
Global scaler	136/256	136/256	136/256	134/256
CS actual	4/31	4/31	8/31	7/31
PWM scale	4	4	8	7
stealthChop	true	true	false	true
msteps		64	64	256	64
tstep		max	max	max	max
PWM thresh.
[mm/s]
OT prewarn	false	false	false	false
triggered
 OTP		false	false	false	false
off time	4	4	0	4
blank time	24	24	36	24
hysteresis
 -end		2	2	-1	2
 -start		1	1	6	1
Stallguard thrs	0	0	0	0
uStep count	1018	1018	894	1018
DRVSTATUS	X	Y	Z	E
sg_result	0	0	0	0
stallguard
fsactive
stst
olb				*
ola				*
s2gb
s2ga
otpw
ot
Driver registers:
		X	0x80:04:40:00
		Y	0x80:04:40:00
		Z	0xE0:08:00:00
		E	0x80:07:40:00
Testing X connection... OK
Testing Y connection... OK
Testing Z connection... OK
Testing E connection... OK

**Comment by ramb0t** [comment]
I have the same problem with SKR Pro V1.2 and BTT TMC5161 V1.1 drivers x5
compiled with latest Marlin 2.0 Bugfix, set rsense to 0.062, 0.060, 0.062f etc. 
all the same, move one or two moves then frozen with no error. Not sure how to debug at this point.

**Comment by Automatica-Mat** [comment]
I ended up swapping to 5160. I have recently thought to try them in stand alone mode. Using the trim pot and jumpers to configure.On Oct 3, 2020 10:03 PM, Rambo <notifications@github.com> wrote:
I have the same problem with SKR Pro V1.2 and BTT TMC5161 V1.1 drivers x5
compiled with latest Marlin 2.0 Bugfix, set rsense to 0.062, 0.060, 0.062f etc.
all the same, move one or two moves then frozen with no error. Not sure how to debug at this point.

**Comment by jimmejames** [comment]
I had good luck with (3) TMC5161's, the latest Marlin (2.0.7.2), BigTreeTech SKR-PRO-V1.1, and CNCJS 1.9.22 on a RPi 3.  I'm currently driving in a dual X's configuration and single stepper for Y- this all works great ((3) NEMA 23's between the X's and Y's) and I'm waiting on a 4th TMC5161 to drive my Z to confirm my configuration.  On previous Marlin builds, I was limited to 5mm/s, only one axis would move at a time, and/or the motor torque was unacceptable, but with this current build, the drives seem stable at 10mm/s, axis move as commanded (e.g. X & Y together), and motors are stronger than I am.  

```
> M122
axis:pwm_scale/curr_scale/mech_load|flags|warncount
        X   X2  Y
Enabled     false   false   false
Set current 1500    1500    1500
RMS current 1482    1482    1482
MAX current 2090    2090    2090
Run current 24/31   24/31   24/31
Hold current    12/31   12/31   12/31
Global scaler   132/256 132/256 132/256
CS actual   12/31   12/31   12/31
PWM scale   11  11  11
vsense
stealthChop true    true    true
msteps      16  16  16
tstep       max max max
PWM thresh.
[mm/s]
OT prewarn  false   false   false
triggered
 OTP        false   false   false
off time    4   4   4
blank time  24  24  24
hysteresis
 -end       2   2   2
 -start     1   1   1
Stallguard thrs 0   0   0
uStep count 8   8   8
DRVSTATUS   X   X2  Y
sg_result   0   0   0
stallguard
fsactive
stst
olb
ola
s2gb
s2ga
otpw
ot
Driver registers:
>       X   0x80:0C:40:00
>       X2  0x80:0C:40:00
        Y   0x80:0C:40:00
Testing X connection... OK
Testing X2 connection... OK
Testing Y connection... OK
ok
```

Here are the configuration files I changed and the firmware output.
[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/5600312/Configuration.zip)

I'm only a few hours of playing into them with this setup, but I've yet to have a hiccup other than after typing M122 into the the CNCJS command line and I get the below repeated until I reset my SKR-PRO.  Problem for a different forum, but included here for completeness.
```
X:20/12/-|TI|   X2:18/12/-|TI|  Y:21/12/-|TI|
X:20/12/-|TI|   X2:18/12/-|TI|  Y:21/12/-|TI|
X:20/12/-|TI|   X2:18/12/-|TI|  Y:21/12/-|TI|
X:20/12/-|TI|   X2:18/12/-|TI|  Y:21/12/-|TI|
```

**Comment by evandene** [comment]
Just my few cents worth of comments:
Have a DELTA printer that needs to home to MAX
My 5x installed TMC5160 do work fine in regard to movements (Latest 2.0x bugfix)
Homing fails, limit switches do show to be triggered (LED's aside the driver and with M119)
My wild guess is that there are no limit endstops PIN's assigned to x, y, and z MAX_PLUG
have my files attached
[DELTA_SKR_PRO_12.zip](https://github.com/MarlinFirmware/Marlin/files/5606947/DELTA_SKR_PRO_12.zip)

**Comment by AndylgTom** [comment]
I have the same problem with SKR1.4turbo and BTT TMC5161 V1.1 drivers.... the motor takes one step and stops, no longer responds. Still not solving the problem?

And everything works fine with TMC5160.

**Comment by evandene** [comment]
Hi
The board was broken and they send me a new one.
When I have time I will setup a Delta Printer with it. For now I'm happy with my Lerdge K board

Ed van den Enden

________________________________

**Comment by thisiskeithb** [comment]
Duplicate of #15105

**Comment by XBrav** [comment]
I would argue this is not closed, and is not a duplicate of #15105 add the device definition. The 5161 issue is not a simple result of a device definition, but rather, the stepper driver will crash if running with the timings for the 5160.

If anything, #15105 should reference this thread for configuration parameters when / if 5161 is ever implemented.

**Comment by thisiskeithb** [comment]
> I would argue this is not closed, and is not a duplicate of #15105 add the device definition. 

One is a duplicate of the other since both are asking for 5161 support. I closed this FR since it was newer of the two, but I’ll open it and close the other since there’s more discussion here.

**Comment by thinkyhead** [comment]
What timings end up working with the 5161?

**Comment by Cailess** [comment]
Do we have any movement on this subject. I have 5 of these that I cannot use. What exactly is this timing mismatch issue were looking at, like what needs to be made?

**Comment by Temtaime** [comment]
I have same issue with FYSETC TMC5160 QHV on latest marlin on Fysetc Spider V2.2.
X and Z motors just stop moving mid print. It happens almost randomly.
Steps become 256 and registers change.
```
Recv: 		X	Y	Z	E
Recv: Enabled		false	false	false	false
Recv: Set current	800	800	400	500
Recv: RMS current	792	792	396	490
Recv: MAX current	1117	1117	558	691
Recv: Run current	15/31	15/31	7/31	9/31
Recv: Hold current	7/31	7/31	0/31	1/31
Recv: Global scaler	133/256	133/256	133/256	133/256
Recv: CS actual	8/31	7/31	8/31	1/31
Recv: PWM scale	8	7	8	1
Recv: vsense
Recv: stealthChop	false	false	false	false
Recv: msteps		256	16	256	16
Recv: interp		true	true	true	true
Recv: tstep		max	max	max	max
Recv: PWM thresh.
Recv: [mm/s]
Recv: OT prewarn	false	false	false	false
Recv: off time	0	5	0	5
Recv: blank time	36	24	36	24
Recv: hysteresis
Recv:  -end		-1	2	-1	2
Recv:  -start		6	4	6	4
Recv: Stallguard thrs	3	3	0	0
Recv: uStep count	1022	872	263	232
Recv: DRVSTATUS	X	Y	Z	E
Recv: sg_result	0	162	0	0
Recv: stallguard				*
Recv: fsactive
Recv: stst
Recv: olb		*		*
Recv: ola		*		*
Recv: s2gb
Recv: s2ga
Recv: otpw
Recv: ot
Recv: Driver registers:
Recv: 		X	0xE0:08:00:00
Recv: 		Y	0x80:07:00:A2
Recv: 		Z	0xE0:08:00:00
Recv: 		E	0x81:01:00:00
Recv: 
Recv: 
Recv: Testing X connection... OK
Recv: Testing Y connection... OK
Recv: Testing Z connection... OK
Recv: Testing E connection... OK
Recv: ok P63 B31
```
Any fix?

**Comment by thisiskeithb** [comment]
@Temtaime: This is a Feature Request for 5161 support, not for posting your issue about 5160s.

You should create a [bug report](https://github.com/MarlinFirmware/Marlin/issues/new/choose) instead of commenting on to unrelated Feature Requests.
